src/Dapr.Client/DaprClientGrpc.cs

```diff
@@ -1373,6 +1375,563 @@ private Autogenerated.StateItem ToAutogeneratedStateItem(StateTransactionRequest
+            }
+        }
+
+        /// <summary>
+        /// Sends the ciphertext bytes in chunks to the sidecar to be decrypted.
+        /// </summary>
+        private async IAsyncEnumerable<byte[]> SendCiphertextStreamAsync(Stream ciphertextStream,
+            int streamingBlockSizeInBytes,
+            AsyncDuplexStreamingCall<Autogenerated.DecryptRequest, Autogenerated.DecryptResponse> duplexStream,
+            Autogenerated.DecryptRequestOptions decryptRequestOptions,
+            [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            //Start with passing the metadata about the decryption request itself in the first message
+            await duplexStream.RequestStream.WriteAsync(
+                new Autogenerated.DecryptRequest { Options = decryptRequestOptions }, cancellationToken);
+            
+            //Send the ciphertext bytes in blocks in subsequent messages
+            await using (var bufferedStream = new BufferedStream(ciphertextStream, streamingBlockSizeInBytes))
+            {
+                var buffer = new byte[streamingBlockSizeInBytes];
+                int bytesRead;
+                ulong sequenceNumber = 0;
+            
+                while ((bytesRead = await bufferedStream.ReadAsync(buffer, 0, streamingBlockSizeInBytes, cancellationToken)) != 0)
+                {
+                    await duplexStream.RequestStream.WriteAsync(new Autogenerated.DecryptRequest
+                    {
+                        Payload = new Autogenerated.StreamPayload
+                        {
+                            Data = ByteString.CopyFrom(buffer, 0, bytesRead),
+                            Seq = sequenceNumber
+                        }
+                    }, cancellationToken);
+            
+                    //Increment the sequence number
+                    sequenceNumber++;
+                    
+                    //Yield an empty byte array to satisfy the return type through it's unused
+                    yield return Array.Empty<byte>();
+                }
+            }
+            
+            //Send the completion message
+            await duplexStream.RequestStream.CompleteAsync();
+            yield break; // End the async enumeration
+        }
+
+        /// <summary>
+        /// Retrieves the decrypted bytes from the decryption operation on the sidecar and returns as an enumerable stream.
+        /// </summary>
+        private async IAsyncEnumerable<byte[]> RetrieveDecryptedStreamAsync(
+            AsyncDuplexStreamingCall<Autogenerated.DecryptRequest, Autogenerated.DecryptResponse> duplexStream,
+            [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            await foreach (var decryptResponse in duplexStream.ResponseStream.ReadAllAsync(cancellationToken)
+                               .ConfigureAwait(false))
+            {
+                yield return decryptResponse.Payload.Data.ToByteArray();
```

Comment:
If the method yielded `ReadOnlyMemory<byte>`, we could eliminate copying the payload data.

Reply:
So, I tried going down this rabbit hole and the short answer is that the ByteString exposed by protobuf only exposes a `ReadOnlySpan<byte>`. This isn't an issue in .NET 7 and up as `Memory<T>` has a new constructor that takes a `Span<T>` as a parameter thus letting us convert to a `ReadOnlyMemory<T>` without copying the payload. Unfortunately, as long as .NET 6 is supported, the only route I'm seeing that works is to copy to a byte array, so while this could yield a `ReadOnlyMemory<byte>`, it's still going to incur the allocation somewhere to get at the bytes in the payload.

I'd propose this be revisited once .NET 6 is EOL'ed later this year.

Reply:
Huh, interesting, as the [docs](https://cloud.google.com/dotnet/docs/reference/Google.Protobuf/latest/Google.Protobuf.ByteString#Google_Protobuf_ByteString_Memory) show `ByteString` having a `ReadOnlyMemory<byte> Memory` property in addition to a `ReadOnlySpan<byte> Span` property.  It's possible that that is the case only in more recent versions of the gRPC NuGet packages, though.

Reply:
It looks like the SDK's existing dependency version has the `Memory` property as well.

Reply:
You're right - I don't know how I missed that yesterday. I've swapped this out to remove the `ToByteArray()` method (and its allocation) to instead point to its `Memory` property, then updated the `EncryptDataAsync` and `DecryptDataAsync` methods to both replace the `List<byte>` buffers with `ArrayBufferWriter<byte>` so it can similarly avoid any array copies.